### DISCUSSION TITLE
High cpu and gpu usage on Windows

### DISCUSSION THREAD

**Discussion by coffee-git** [slint-ui/slint#3445] [General] [answered]
![Screenshot 2023-09-11 181521](https://github.com/slint-ui/slint/assets/71521983/e8130b64-4f36-4d76-b154-6e318817524b)
Hi! Tried to run examples, why is CPU and GPU usage so high? Windows 11, rustc 1.72.0, compiled with release profile.

**Comment by ogoffart** [comment]
Do you know what slint backend and renderer are you using? (Is it the default? Do you have Qt?)

The main page of the gallery has a progress indicator that animates. Do you see the same usage when not showing the animated progress indicator?

**Reply by coffee-git** [reply]
Just cargo build --release, all other settings are default. Todo example when i just move mouse over the window. When i start adding new entries, about 30 percent of CPU usage.
![image](https://github.com/slint-ui/slint/assets/71521983/0ada5c82-d18e-4155-afbb-803e8d8214be)

**Reply by coffee-git** [reply]
Better with skia renderer, but the gpu load still high

**Comment by tronical** [comment]
The gallery runs an animation and I think that we don't throttle it (on Windows). So we try to render as fast as we can, possibly at a rate higher than the screen refresh rate. That would result in a CPU and GPU load that's higher than necessary to achieve the desired animation.

**Reply by coffee-git** [reply]
On skia there are fixed 60 frames, in the gallery example the load on the GPU is consistently 15%, in todo it peaks up to 10% when I just move the mouse, isn’t it too much?

![image](https://github.com/slint-ui/slint/assets/71521983/2a4a118e-af45-45e0-a910-1b298202ae81)


```
Slint: Build config: release; Backend: FemtoVG renderer
average frames per second: 2 details from last frame: [0 layers created]
average frames per second: 121 details from last frame: [0 layers created]
average frames per second: 35 details from last frame: [0 layers created]
average frames per second: 507 details from last frame: [0 layers created]
average frames per second: 510 details from last frame: [0 layers created]
average frames per second: 470 details from last frame: [0 layers created]
average frames per second: 497 details from last frame: [0 layers created]
average frames per second: 212 details from last frame: [0 layers created]
average frames per second: 0
average frames per second: 248 details from last frame: [0 layers created]
average frames per second: 505 details from last frame: [0 layers created]
average frames per second: 116 details from last frame: [0 layers created]
```

```
Slint: Build config: release; Backend: Skia renderer (skia backend d3d; surface: 32 bpp)
average frames per second: 28
average frames per second: 0
average frames per second: 46
average frames per second: 60
average frames per second: 44
average frames per second: 21
average frames per second: 55
average frames per second: 60
average frames per second: 61
average frames per second: 26
average frames per second: 62
average frames per second: 60
average frames per second: 60
average frames per second: 60
```

**Reply by tronical** [reply]
Ahh yes, with Direct3D the swap chain seems throttled. I see the same on my machine. That's good. In terms of CPU and GPU usage, I see roughly the same on a 2019 MBP and an Intel UHD Graphics 630.

Note that we don't support partial rendering yet, so this is rendering the entire window every time.

What's your target percentage number?

**Reply by coffee-git** [reply]
Just realized that most gui application uses about the same amount of resources. Didn't pay attention to it before.
Is it possible to limit the number of frames for FemtoVG?

**Reply by ogoffart** [reply]
> Is it possible to limit the number of frames

IIRC this will be a feature coming with winit 0.29

**Reply by tronical** [reply]
I think there's a bug on our side here. We default to letting the OpenGL driver pick a swap interval, which happens to be correct for EGL. But it looks like on Windows it's zero, while we should really default to vsync.

Winit's upcoming frame throttling is a potential alternative, but unfortunately it's only implemented for Wayland at the moment.

**Reply by tronical** [reply, marked as answer]
I created https://github.com/slint-ui/slint/issues/3516 to track the vsync setting.

Edit: This is fixed now :)
